libminifi/src/core/ProcessSession.cpp

```diff
@@ -212,6 +212,10 @@ void ProcessSession::transfer(const std::shared_ptr<core::FlowFile> &flow, Relat
 }
 
 void ProcessSession::write(const std::shared_ptr<core::FlowFile> &flow, OutputStreamCallback *callback) {
+  gsl_ExpectsAudit(_updatedFlowFiles.contains(flow->getUUID())
```

Comment:
as we provide access to the session to custom code through the script extension, this feels like an invariant we should always check, turning the `std::map`s (and the `_clonedFlowFiles` vector) into `std::unordered_map`s would make the check an `O(1)` operation

Reply:
I prefer the narrow contract, but if you decide to have a wide contract, then please use if + throw or something similar to implement it, not gsl_Expects (which is meant to declare preconditions).

Reply:
we wrap the session in a `PyProcessSession` which could handle the check if it had access to the internals of the `ProcessSession` (not probable) or had its own flowfile registry (would be duplicating the session's internal registry)

Reply:
I would not be against a public member function that can check if a flow file can be written to. I don't know a good name though. Maybe `bool canBeWrittenTo(const core::FlowFile&) const noexcept;`? If we did that, `PyProcessSession` could check that and `write`/`append` could have that as their precondition.